### ISSUE TITLE
--bisect and --exclude-pattern options do not play well together

### ISSUE THREAD

**Issue by geeksam** [rspec/rspec-core#2122]
The following command line (I swiped the ```--pattern``` option from Rake output, and wrote the ```--exclude-pattern``` myself):

```
rspec --pattern spec/\*\*\{,/\*/\*\ *\}/\*_spec.rb --exclude-pattern './spec/features/**/*_spec.rb' --order 28534 --bisect
```

...ran for six hours(!), eventually producing the following final output:

```
Bisect complete! Reduced necessary non-failing examples from 698 to 4 in 342 minutes 49 seconds.

The minimal reproduction command is:
  rspec ./spec/features/foo_spec.rb[1:1:1,1:1:2,1:2:1,1:2:2,1:2:3:1,1:2:3:2,1:2:3:3,1:2:4:1,1:2:4:2,1:3:1,1:3:2] ./spec/features/bar_spec.rb[1:1:1:1,1:1:2:1] ./spec/models/lead_spec.rb[2:2:2] ./spec/models/yak_spec.rb[1:1:1:1,1:1:3:1,1:2:1:2,1:3:1:1,1:3:1:2] ./spec/models/shave_spec.rb[1:5:1,1:5:3,1:7:2,1:11:2,1:11:3] --pattern spec/**{,/*/**}/*_spec.rb --exclude-pattern ./spec/features/**/*_spec.rb --order 28534
```

In this particular app on my current machine, ./spec/features/*_spec.rb tends to be (a) slow and (b) flaky, which is why I had tried to skip them using the ```--exclude-pattern``` option in the first place.  Indeed, the first time I ran the minimal repro command, it hung.  Out of curiosity, I removed the two feature specs and reran the command, and got a failure.

Upon closer inspection, the identifiers in the "minimal reproduction command" for both files in ./spec/features are just a verbose way of including every example in those files, which I find suspicious as well.  (The identifiers for the files in ./spec/models are both proper subsets of the available examples, so I have more confidence that bisect has done its job there.)

Is the ```--bisect``` option:

  1. ignoring the ```--exclude-pattern``` option and including all examples in its run,
  2. paying attention to ```--exclude-pattern``` but then mistakenly including excluded examples when building the minimal repro command, or
  3. other?

I might try deleting the ./spec/features directory and running another bisect to attempt to distinguish option 1 from option 2, but that'll have to wait until the end of my workday.

**Comment by myronmarston** [comment]
I'm not quote sure what it's doing (I'd have to dig in, because the behavior you described is puzzling), but I have a theory: when bisect runs, it tries to use your command line arguments but with the [file and location arguments stripped out](https://github.com/rspec/rspec-core/blob/v3.4.1/lib/rspec/core/bisect/runner.rb#L93) -- the idea being that if you do `rspec spec/unit --bisect` it needs to remove `spec/unit` from the individual runs it does so it can pick specific examples to run each time.  If it merely appended its example selections to the command it runs, it would run all specs in `spec/unit` plus the specific examples it intends to run.  My guess is that `--exclude-pattern` is being wrongly categorized as a location argument and being stripped out.

Have you tried running with `--bisect=verbose`?  It'll print off each command that it runs which should hopefully explain what's happening.

https://relishapp.com/rspec/rspec-core/v/3-4/docs/command-line/bisect#use-`--bisect=verbose`-to-enable-verbose-debug-mode-for-more-detail

**Comment by geeksam** [comment]
Sorry for delay in replying; was away for holidays. :)

TL;DR: your guess seems reasonable.

---

The full suite was way too slow, so I added a few bogus tests ('foo', 'bar', 'yak', and 'maybe?') in ```./spec/models/foo_spec.rb``` with "foo: true" in their metadata, then set them up to fail if run in a particular order (specifically, 'maybe?' fails if run immediately after 'bar'). I also added a spec in ```./spec/features/foo_spec.rb``` that would *always* fail.

Here's the output from a failing test run with ```--format documentation```:

```
$ bundle exec rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb --exclude-pattern './spec/features/**/*_spec.rb' -t foo --seed 43231 --format documentation
I, [2015-11-30T11:40:55.502137 #21669]  INFO -- : Not reporting to Code Climate because ENV['CODECLIMATE_REPO_TOKEN'] is not set.
/Users/sam/.gem/ruby/2.2.0/gems/fog-1.14.0/lib/fog/aws/compute.rb:200: warning: duplicated key at line 208 ignored: :internet_gateways
Digest::Digest is deprecated; use Digest
/Users/sam/.gem/ruby/2.2.0/gems/fog-1.14.0/lib/fog/rackspace/mock_data.rb:41: warning: duplicated key at line 79 ignored: "name"
You are using Excon 0.25.3. WebMock supports version >= 0.27.5.
I, [2015-11-30T11:40:59.648562 #21669]  INFO -- : ** [Raven] Raven 0.12.3 ready to catch errors
Run options: include {:focus=>true, :foo=>true}

Randomized with seed 43231

Foo
  foo
  yak
  bar
  maybe? (FAILED - 1)

Failures:

  1) Foo maybe?
     Failure/Error: fail "nope" if $fail

     RuntimeError:
       nope
     # ./spec/models/foo_spec.rb:17:in `block (2 levels) in <top (required)>'

##### [Profiling output skipped -SLG] #####

Finished in 0.26425 seconds (files took 5.66 seconds to load)
4 examples, 1 failure

Failed examples:

rspec ./spec/models/foo_spec.rb:16 # Foo maybe?

Randomized with seed 43231
```

Once that was confirmed, I replaced ```--format documentation``` with ```--bisect=verbose```:

```
$ bundle exec rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb --exclude-pattern './spec/features/**/*_spec.rb' -t foo --seed 43231 --bisect=verbose
Bisect started using options: "--pattern spec/**{,/*/**}/*_spec.rb --exclude-pattern ./spec/features/**/*_spec.rb -t foo --seed 43231"
Running suite to find failures... (6.96 seconds)
 - Failing examples (2):
    - ./spec/features/foo_spec.rb[1:1]
    - ./spec/models/foo_spec.rb[1:4]
 - Non-failing examples (3):
    - ./spec/models/foo_spec.rb[1:1,1:2,1:3]
Checking that failure(s) are order-dependent..
 - Running: rspec ./spec/features/foo_spec.rb[1:1] ./spec/models/foo_spec.rb[1:4] --pattern spec/**{,/*/**}/*_spec.rb --exclude-pattern ./spec/features/**/*_spec.rb -t foo --seed 43231 (7.08 seconds)
 - Failure appears to be order-dependent
Round 1: bisecting over non-failing examples 1-3
 - Running: rspec ./spec/features/foo_spec.rb[1:1] ./spec/models/foo_spec.rb[1:2,1:4] --pattern spec/**{,/*/**}/*_spec.rb --exclude-pattern ./spec/features/**/*_spec.rb -t foo --seed 43231 (7.17 seconds)
 - Examples we can safely ignore (2):
    - ./spec/models/foo_spec.rb[1:1,1:3]
 - Remaining non-failing examples (1):
    - ./spec/models/foo_spec.rb[1:2]
Bisect complete! Reduced necessary non-failing examples from 3 to 1 in 14.25 seconds.

The minimal reproduction command is:
  rspec ./spec/features/foo_spec.rb[1:1] ./spec/models/foo_spec.rb[1:2,1:4] --pattern spec/**{,/*/**}/*_spec.rb --exclude-pattern ./spec/features/**/*_spec.rb -t foo --seed 43231
```

Note that ```./spec/features/foo_spec.rb``` is included in bisect's output, which at least does not disprove your hypothesis.  ;)

**Comment by myronmarston** [comment]
I can't repro.  I put my repro attempt in https://github.com/myronmarston/rspec-bisect-issue-2122.  Here's the output w/o the `--exclude-pattern` option:

```
Bisect started using options: "--seed 1234"
Running suite to find failures... (0.21552 seconds)
 - Failing examples (1):
    - './spec/calculator_1_spec.rb[1:1]'
 - Non-failing examples (10):
    - './spec/calculator_10_spec.rb[1:1]'
    - './spec/calculator_2_spec.rb[1:1]'
    - './spec/calculator_3_spec.rb[1:1]'
    - './spec/calculator_4_spec.rb[1:1]'
    - './spec/calculator_5_spec.rb[1:1]'
    - './spec/calculator_6_spec.rb[1:1]'
    - './spec/calculator_7_spec.rb[1:1]'
    - './spec/calculator_8_spec.rb[1:1]'
    - './spec/calculator_9_spec.rb[1:1]'
    - './spec/features/some_slow_spec.rb[1:1]'

Round 1: searching for 5 non-failing examples (of 10) to ignore:
 - Running: rspec './spec/calculator_1_spec.rb[1:1]' './spec/calculator_4_spec.rb[1:1]' './spec/calculator_6_spec.rb[1:1]' './spec/calculator_7_spec.rb[1:1]' './spec/calculator_8_spec.rb[1:1]' './spec/calculator_9_spec.rb[1:1]' --seed 1234 (0.1971 seconds)
 - Running: rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' './spec/calculator_2_spec.rb[1:1]' './spec/calculator_3_spec.rb[1:1]' './spec/calculator_5_spec.rb[1:1]' './spec/features/some_slow_spec.rb[1:1]' --seed 1234 (0.22007 seconds)
 - Examples we can safely ignore (5):
    - './spec/calculator_4_spec.rb[1:1]'
    - './spec/calculator_6_spec.rb[1:1]'
    - './spec/calculator_7_spec.rb[1:1]'
    - './spec/calculator_8_spec.rb[1:1]'
    - './spec/calculator_9_spec.rb[1:1]'
 - Remaining non-failing examples (5):
    - './spec/calculator_10_spec.rb[1:1]'
    - './spec/calculator_2_spec.rb[1:1]'
    - './spec/calculator_3_spec.rb[1:1]'
    - './spec/calculator_5_spec.rb[1:1]'
    - './spec/features/some_slow_spec.rb[1:1]'
 - Round finished (0.41817 seconds)
Round 2: searching for 3 non-failing examples (of 5) to ignore:
 - Running: rspec './spec/calculator_1_spec.rb[1:1]' './spec/calculator_2_spec.rb[1:1]' './spec/calculator_5_spec.rb[1:1]' --seed 1234 (0.22315 seconds)
 - Running: rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' './spec/calculator_3_spec.rb[1:1]' './spec/features/some_slow_spec.rb[1:1]' --seed 1234 (0.23197 seconds)
 - Examples we can safely ignore (2):
    - './spec/calculator_2_spec.rb[1:1]'
    - './spec/calculator_5_spec.rb[1:1]'
 - Remaining non-failing examples (3):
    - './spec/calculator_10_spec.rb[1:1]'
    - './spec/calculator_3_spec.rb[1:1]'
    - './spec/features/some_slow_spec.rb[1:1]'
 - Round finished (0.45597 seconds)
Round 3: searching for 2 non-failing examples (of 3) to ignore:
 - Running: rspec './spec/calculator_1_spec.rb[1:1]' './spec/features/some_slow_spec.rb[1:1]' --seed 1234 (0.19694 seconds)
 - Running: rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' './spec/calculator_3_spec.rb[1:1]' --seed 1234 (0.21247 seconds)
 - Examples we can safely ignore (1):
    - './spec/features/some_slow_spec.rb[1:1]'
 - Remaining non-failing examples (2):
    - './spec/calculator_10_spec.rb[1:1]'
    - './spec/calculator_3_spec.rb[1:1]'
 - Round finished (0.41015 seconds)
Round 4: searching for 1 non-failing example (of 2) to ignore:
 - Running: rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' --seed 1234 (0.20396 seconds)
 - Examples we can safely ignore (1):
    - './spec/calculator_3_spec.rb[1:1]'
 - Remaining non-failing examples (1):
    - './spec/calculator_10_spec.rb[1:1]'
 - Round finished (0.20442 seconds)
Round 5: searching for 1 non-failing example (of 1) to ignore:
 - Running: rspec './spec/calculator_1_spec.rb[1:1]' --seed 1234 (0.2112 seconds)
 - Round finished (0.21147 seconds)
Bisect complete! Reduced necessary non-failing examples from 10 to 1 in 1.7 seconds.

The minimal reproduction command is:
  rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' --seed 1234
```

Notice that it includes `./spec/features/some_slow_spec.rb`.  Running with the `--exclude-pattern` option:

```
Bisect started using options: "--seed 1234 --exclude-pattern spec/features/**/*_spec.rb"
Running suite to find failures... (0.21813 seconds)
 - Failing examples (1):
    - './spec/calculator_1_spec.rb[1:1]'
 - Non-failing examples (9):
    - './spec/calculator_10_spec.rb[1:1]'
    - './spec/calculator_2_spec.rb[1:1]'
    - './spec/calculator_3_spec.rb[1:1]'
    - './spec/calculator_4_spec.rb[1:1]'
    - './spec/calculator_5_spec.rb[1:1]'
    - './spec/calculator_6_spec.rb[1:1]'
    - './spec/calculator_7_spec.rb[1:1]'
    - './spec/calculator_8_spec.rb[1:1]'
    - './spec/calculator_9_spec.rb[1:1]'

Round 1: searching for 5 non-failing examples (of 9) to ignore:
 - Running: rspec './spec/calculator_1_spec.rb[1:1]' './spec/calculator_6_spec.rb[1:1]' './spec/calculator_7_spec.rb[1:1]' './spec/calculator_8_spec.rb[1:1]' './spec/calculator_9_spec.rb[1:1]' --seed 1234 --exclude-pattern spec/features/**/*_spec.rb (0.20601 seconds)
 - Running: rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' './spec/calculator_2_spec.rb[1:1]' './spec/calculator_3_spec.rb[1:1]' './spec/calculator_4_spec.rb[1:1]' './spec/calculator_5_spec.rb[1:1]' --seed 1234 --exclude-pattern spec/features/**/*_spec.rb (0.21536 seconds)
 - Examples we can safely ignore (4):
    - './spec/calculator_6_spec.rb[1:1]'
    - './spec/calculator_7_spec.rb[1:1]'
    - './spec/calculator_8_spec.rb[1:1]'
    - './spec/calculator_9_spec.rb[1:1]'
 - Remaining non-failing examples (5):
    - './spec/calculator_10_spec.rb[1:1]'
    - './spec/calculator_2_spec.rb[1:1]'
    - './spec/calculator_3_spec.rb[1:1]'
    - './spec/calculator_4_spec.rb[1:1]'
    - './spec/calculator_5_spec.rb[1:1]'
 - Round finished (0.4224 seconds)
Round 2: searching for 3 non-failing examples (of 5) to ignore:
 - Running: rspec './spec/calculator_1_spec.rb[1:1]' './spec/calculator_4_spec.rb[1:1]' './spec/calculator_5_spec.rb[1:1]' --seed 1234 --exclude-pattern spec/features/**/*_spec.rb (0.21419 seconds)
 - Running: rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' './spec/calculator_2_spec.rb[1:1]' './spec/calculator_3_spec.rb[1:1]' --seed 1234 --exclude-pattern spec/features/**/*_spec.rb (0.21411 seconds)
 - Examples we can safely ignore (2):
    - './spec/calculator_4_spec.rb[1:1]'
    - './spec/calculator_5_spec.rb[1:1]'
 - Remaining non-failing examples (3):
    - './spec/calculator_10_spec.rb[1:1]'
    - './spec/calculator_2_spec.rb[1:1]'
    - './spec/calculator_3_spec.rb[1:1]'
 - Round finished (0.42914 seconds)
Round 3: searching for 2 non-failing examples (of 3) to ignore:
 - Running: rspec './spec/calculator_1_spec.rb[1:1]' './spec/calculator_2_spec.rb[1:1]' --seed 1234 --exclude-pattern spec/features/**/*_spec.rb (0.19796 seconds)
 - Running: rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' './spec/calculator_3_spec.rb[1:1]' --seed 1234 --exclude-pattern spec/features/**/*_spec.rb (0.20947 seconds)
 - Examples we can safely ignore (1):
    - './spec/calculator_2_spec.rb[1:1]'
 - Remaining non-failing examples (2):
    - './spec/calculator_10_spec.rb[1:1]'
    - './spec/calculator_3_spec.rb[1:1]'
 - Round finished (0.40815 seconds)
Round 4: searching for 1 non-failing example (of 2) to ignore:
 - Running: rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' --seed 1234 --exclude-pattern spec/features/**/*_spec.rb (0.20641 seconds)
 - Examples we can safely ignore (1):
    - './spec/calculator_3_spec.rb[1:1]'
 - Remaining non-failing examples (1):
    - './spec/calculator_10_spec.rb[1:1]'
 - Round finished (0.20678 seconds)
Round 5: searching for 1 non-failing example (of 1) to ignore:
 - Running: rspec './spec/calculator_1_spec.rb[1:1]' --seed 1234 --exclude-pattern spec/features/**/*_spec.rb (0.21017 seconds)
 - Round finished (0.21043 seconds)
Bisect complete! Reduced necessary non-failing examples from 9 to 1 in 1.68 seconds.

The minimal reproduction command is:
  rspec './spec/calculator_10_spec.rb[1:1]' './spec/calculator_1_spec.rb[1:1]' --seed 1234 --exclude-pattern spec/features/**/*_spec.rb
```

Notice that `./spec/features/some_slow_spec.rb` is correctly excluded.

So apparently `--bisect` and `--exclude-pattern` _can_ generally play together but there's something different in your case that I'm not able to guess.  I'm going to close this one as it's currently unactionable without a repro case, but if you're able to provide a repro case, @geeksam, we can re-open and look into it some more.
